### PR TITLE
Support for AddOn parameters

### LINKED ISSUE #3380: Support ODF Addon configuration in ibm_container_addons
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Today, `ibm_container_addons` supports ODF but it comes with default values.
```
monStorageClassName   ibmc-vpc-block-metro-10iops-tier   
osdSize               250Gi   
osdDevicePaths        invalid   
workerNodes           all   
ocsUpgrade            false   
odfDeploy             true   
monSize               20Gi   
numOfOsd              1   
monDevicePaths        invalid   
osdStorageClassName   ibmc-vpc-block-metro-10iops-tier
clusterEncryption     false
```
The terraform provider should allow to configure the parameter values above or any other parameters that will be introduced in the future. That will make it on par with the UI and the CLI.

```
ibmcloud ks cluster addon options --addon openshift-data-foundation

Add-on Options
Option                     Default Value
monSize                    20Gi
monDevicePaths             <Please provide IDs of the disks to be used for mon pods if using local disks or standard classic cluster>
osdStorageClassName        ibmc-vpc-block-metro-10iops-tier
workerNodes                all
clusterEncryption          false
billingType                advanced
containerPrivateEndpoint   <Please provide the satellite link for the container private endpoint if using a satellite cluster>
odfDeploy                  true
monStorageClassName        ibmc-vpc-block-metro-10iops-tier
osdSize                    250Gi
osdDevicePaths             <Please provide IDs of the disks to be used for OSD pods if using local disks or standard classic cluster>
numOfOsd                   1
ocsUpgrade                 false
```

Terraform Automation is required to install ODF but also Cloud Paks.

### New or Affected Resource(s)

* ibm_container_addons

### PR DESCRIPTION
**Pull request by aayushsss1**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #3380

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 1]
```diff
@@ -551,3 +735,30 @@ func resourceIBMContainerAddonsHash(v interface{}) int {
 
 	return conns.String(buf.String())
 }
+
+// Check if correct parameters are provided by the user
+func validateAddonOptions(jsonInput map[string]interface{}, yamlDefault map[string]interface{}) (bool, error) {
```
Comment (hasan4791) [inline ibm/service/kubernetes/resource_ibm_container_addons.go:740]: pass the inputs as pointers.

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 2]
```diff
@@ -551,3 +735,30 @@ func resourceIBMContainerAddonsHash(v interface{}) int {
...
+		}
+
+	}
+
+	return true, nil
+}
+
+// Set the Value of the Parameters provided by the user in terraform in the config map through index matching
+
+func updateAddOnOptions(jsonInput map[string]interface{}, yamlDefault map[string]interface{}) map[string]interface{} {
```
Comment (hasan4791) [inline ibm/service/kubernetes/resource_ibm_container_addons.go:756]: pass the inputs as pointers & we can remove the return statement from this function.

File: website/docs/d/container_addons.html.markdown [hunk 3]
```diff
@@ -19,10 +19,33 @@ data "ibm_container_addons" "addons" {
...
+}
+ ```
+
 ## Argument reference
 Review the argument references that you can specify for your data source. 
 
 - `cluster` - (Required, String) The name or ID of the cluster.
+- `manage_all_addons` - (Optional, Bool) To manage all add-ons installed in the cluster using terraform by importing it into the state file, default is set to `true`.
 
 ## Attribute reference
```
Comment (hasan4791) [inline website/docs/d/container_addons.html.markdown:50]: Add "options" parameter also.

**Review by hasan4791 — CHANGES_REQUESTED** [review]

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 4]
```diff
@@ -390,7 +572,7 @@ func resourceIBMContainerAddOnsUpdate(d *schema.ResourceData, meta interface{})
 							Name: newPack["name"].(string),
 						}
 						if newPack["version"] != nil {
-							rmParam.Version = newPack["version"].(string)
```
[the remark this answers is not in the document]
Comment (aayushsss1) [reply ibm/service/kubernetes/resource_ibm_container_addons.go:393]: Yeah, I was going through the code and uncovered it

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 5]
```diff
@@ -166,8 +198,75 @@ func resourceIBMContainerAddOnsCreate(d *schema.ResourceData, meta interface{})
...
+func getOptions(addOn map[string]interface{}, d *schema.ResourceData, meta interface{}) (string, error) {
+
+	csClient, err := meta.(conns.ClientSession).ContainerAPI()
+	if err != nil {
+		return "", err
+	}
+
+	addOnAPI := csClient.AddOns()
+
+	// Check if parameters are given in terraform, if no, then return
```
[the remark this answers is not in the document]
Comment (aayushsss1) [reply ibm/service/kubernetes/resource_ibm_container_addons.go:212]: csClient, err := meta.(conns.ClientSession).ContainerAPI(), is above this line, so I've defined this just below

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 6]
```diff
@@ -113,6 +121,25 @@ func ResourceIBMContainerAddOns() *schema.Resource {
 							Computed:    true,
 							Description: "VLAN spanning required for multi-zone clusters",
 						},
+						"options": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							ForceNew:    false,
```
[the remark this answers is not in the document]
Comment (aayushsss1) [reply ibm/service/kubernetes/resource_ibm_container_addons.go:127]: Which parameter? ForceNew in options?

File: examples/openshift-data-foundation/deployment/README.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
+# Deploying and Managing Openshift Data Foundation (Test)
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:1]: change the title to 
```
[Tech Preview] Deploying and Managing Openshift Data Foundation
```

File: examples/openshift-data-foundation/deployment/README.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+This example shows how to Deploy and Manage your Openshift Data Foundation add-on on a Red Hat Openshift Cluster. Note this template is still in development, so please be advised before using in production.
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:3]: Change this line as 
```
This example shows how to deploy and manage the Openshift Data Foundation (ODF) on IBM Cloud VPC based RedHat Openshift cluster.
```

File: examples/openshift-data-foundation/deployment/README.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+This sample configuration will deploy the add-on, scale and upgrade it using the "ibm_container_addons" and kubernetes_manifest from the ibm terraform provider and kubernetes provider respectively.
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:5]: change this line 
```
This sample configuration will deploy the add-on
```
as below
```
This sample configuration will deploy the ODF
```

File: examples/openshift-data-foundation/deployment/README.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+$ terraform init
+$ terraform plan --var-file input.tfvars
+$ terraform apply --var-file input.tfvars
+```
+
+Run `terraform destroy` when you don't need these resources.
+
+## Example usage
+
+The following use cases are handled by editing the input.tfvars file. Terraform resources need to be touched.
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:21]: Remove the line `Terraform resources need to be touched.`

File: examples/openshift-data-foundation/deployment/README.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+### Create & Deployment:
+
+The default input.tfvars is given below, the user should just change the value of the parameters in accorandance to their requirment. 
+
+```hcl
+# For add-on deployment
+odfVersion = "4.10.0"
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:29]: Change default odfVersion to `4.12`

File: examples/openshift-data-foundation/deployment/README.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+$ terraform apply --var-file input.tfvars
+```
+
+Run `terraform destroy` when you don't need these resources.
+
+## Example usage
+
+The following use cases are handled by editing the input.tfvars file. Terraform resources need to be touched.
+
+### Create & Deployment:
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:23]: Change this to `ODF deployment`

File: examples/openshift-data-foundation/deployment/README.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+hpcsTokenUrl = null
+ignoreNoobaa = "false"
+numOfOsd = "1"
+ocsUpgrade = "false"
+osdDevicePaths = null
+osdSize = "100Gi"
+osdStorageClassName = "ibmc-vpc-block-metro-10iops-tier"
+workerNodes = null
+```
+### Editing the Ocscluster custom resource
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:50]: change this to `ODF scale-up`

File: examples/openshift-data-foundation/deployment/README.md [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+The following variables in the `input.tfvars` file can be edited
+
+* numOfOsd - To scale your storage
+* workerNodes - To increase the number of workerNodes with ODF
+
+```hcl
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:57]: Remove `hcl` if not needed

File: examples/openshift-data-foundation/deployment/README.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+# For CRD Management
+numOfOsd = "1" -> "2"
+workerNodes = null -> "worker_1_ID,worker_2_ID"
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:60]: change the example to 
```
workerNodes = "worker_1_ID,worker_2_ID,worker_1_ID,worker_3_ID" -> "worker_1_ID,worker_2_ID,worker_1_ID,worker_3_ID,worker_4_ID,worker_5_ID,worker_1_ID,worker_6_ID"
```
Also add a note for workerNodes that it should be updated only if workerNodes was provided as input during ODF deploy.

File: examples/openshift-data-foundation/deployment/README.md [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+```
+
+### Update
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:63]: change this to `ODF upgrade`

File: examples/openshift-data-foundation/deployment/README.md [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+
+The following variables in the `input.tfvars` file should be changed in order to upgrade the ODF add-on and the Ocscluster CRD.
+
+* odfVersion - Specify the version you wish to upgrade to
+* ocsUpgrade - Must be set to `true` to upgrade the CRD 
+
+```hcl
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:70]: Remove `hcl`

File: examples/openshift-data-foundation/deployment/README.md [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+# For ODF add-on upgrade
+odfVersion = "4.10.0" -> "4.11.0"
+
+# For Ocscluster upgrade
+workerNodes = "false" -> "true"
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:75]: I think it should be `ocsUpgrade`

File: examples/openshift-data-foundation/deployment/README.md [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+### Update
+
+The following variables in the `input.tfvars` file should be changed in order to upgrade the ODF add-on and the Ocscluster CRD.
+
+* odfVersion - Specify the version you wish to upgrade to
+* ocsUpgrade - Must be set to `true` to upgrade the CRD 
+
+```hcl
+# For ODF add-on upgrade
+odfVersion = "4.10.0" -> "4.11.0"
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/README.md:72]: change version from 4.12 to 4.13

File: examples/openshift-data-foundation/deployment/input.tfvars [hunk 20]
```diff
@@ -0,0 +1,31 @@
+## DEFAULT VALUES ARE SET ##
+## Please change according to your configuratiom ##
+
+
+# To enable ODF AddOn on your cluster
+ibmcloud_api_key = ""
+cluster = ""
+region = "us-south"
+odfVersion = "4.10.0"
```
Comment (akgunjal) [inline examples/openshift-data-foundation/deployment/input.tfvars:9]: change it to 4.12.0

**Review by akgunjal — COMMENTED** [review]
common comments are given here:

1. The folder structure should be `examples/openshift-data-foundation/addon/main.tf`
2. Change the `AddOn` folder to `ibm_odf_addon`
3. Change the `CRD` folder to `ocscluster`
4. Also change shell script file names to all lowercase

File: examples/openshift-data-foundation/deployment/README.md [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+## Example usage
+
+The following use cases are handled by editing the input.tfvars file. Terraform resources need to be touched.
+
+### Create & Deployment:
+
+The default input.tfvars is given below, the user should just change the value of the parameters in accorandance to their requirment. 
+
+```hcl
```
[the remark this answers is not in the document]
Comment (aayushsss1) [reply examples/openshift-data-foundation/deployment/README.md:27]: configuration language, it's used to depict the terraform files in readme

File: examples/openshift-data-foundation/deployment/README.md [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+# For ODF add-on upgrade
+odfVersion = "4.10.0" -> "4.11.0"
+
+# For Ocscluster upgrade
+workerNodes = "false" -> "true"
+```
+
+## Examples
+
+* [ ODF Deployment & Management ](https://github.com/IBM-Cloud/terraform-provider-ibm/tree/master/examples/openshift-data-foundation/deployment)
```
[the remark this answers is not in the document]
Comment (aayushsss1) [reply examples/openshift-data-foundation/deployment/README.md:80]: The other examples were similar so I added this, I'm not sure if it's required. Will check and make the changes

**Comment by hasan4791** [comment]
@hkantare @kavya498 
Could you please take a look at this PR!!

**Review by hkantare — APPROVED** [review]

**Pull request merged**